compiler/src/dotty/tools/dotc/transform/Pickler.scala

```diff
@@ -85,8 +95,20 @@ class Pickler extends Phase {
     do
       if ctx.settings.YtestPickler.value then beforePickling(cls) = tree.show
 
+      val isJavaAttr =
+        // we must always set JAVAattr when pickling Java sources,
+        // additionally assert that Java sources didn't reach Pickler without `-Yjava-tasty`.
+        unit.isJava && true.ensuring(ctx.settings.YjavaTasty.value, "unexpected Java source file without -Yjava-tasty")
```

Comment:
This seems like a convoluted way of doing the following:
```suggestion
      val isJavaAttr = unit.isJava // we must always set JAVAattr when pickling Java sources
      // assert that Java sources didn't reach Pickler without `-Yjava-tasty`.
      if isJavaAttr then
        assert(ctx.settings.YjavaTasty.value, "unexpected Java source file without -Yjava-tasty")

Reply:
fixed